src/xcode/ENA/ENA/Source/Scenes/ExposureDetection/ExposureDetectionViewController.swift

```diff
@@ -200,8 +204,51 @@ extension ExposureDetectionViewController {
 			checkButton.isEnabled = true
 		case .manual:
 			footerView.isHidden = false
-			checkButton.setTitle(AppStrings.ExposureDetection.buttonRefresh, for: .normal)
-			checkButton.isEnabled = riskProvider.manualExposureDetectionState == .possible
+
+			let nextRefresh = riskProvider.nextExposureDetectionDate()
+			let now = Date()
+
+			if nextRefresh > now, countdown == nil {
+				scheduleCountdownTimer(to: nextRefresh)
+			} else if let countdown = countdown, countdown.end != nextRefresh {
+				// Make sure to schedule new countdown if the next refresh time has changed.
+				scheduleCountdownTimer(to: nextRefresh)
+			} else if nextRefresh - 1 > now {
+				showCountdownButton(with: time)
+			} else {
+				showActiveRefreshButton()
+			}
 		}
 	}
+
+	private func scheduleCountdownTimer(to end: Date) {
+		countdown = CountdownTimer(countdownTo: end)
+		countdown?.delegate = self
+		countdown?.start()
+	}
+
+	private func showActiveRefreshButton() {
+		self.checkButton.setTitle(AppStrings.ExposureDetection.buttonRefresh, for: .normal)
+		self.checkButton.isEnabled = riskProvider.manualExposureDetectionState == .possible
+	}
+
+	private func showCountdownButton(with time: String? = nil) {
+		guard let time = time else { return }
+		UIView.performWithoutAnimation {
+			self.checkButton.setTitle(String(format: AppStrings.ExposureDetection.refreshIn, time), for: .normal)
+			self.checkButton.isEnabled = riskProvider.manualExposureDetectionState == .possible
+		}
+	}
+}
+
+// MARK: CountdownTimerDelegate methods.
+
+extension ExposureDetectionViewController: CountdownTimerDelegate {
+	func update(time: String) {
+		self.updateCheckButton(time)
+	}
+
+	func done() {
+		self.updateCheckButton()
```

Comment:
Perhaps just making an `invalidateCheckButton()` instead of using the `updateCheckButton` would be more clean here. Then you also don't need to check in `updateCheckButton` whether time is nil or not. Also would be more self-explanatory imho.

Reply:
i agree with your concerns, however the current way is safer when it comes to changing parameters, e.g. a user switching on automatic mode. I added comments to the updateCheckButton method to make the flow a bit more clear